### DISCUSSION TITLE
Clarification for Power calibration feature on RCP side.

### DISCUSSION THREAD

**Discussion by aashu216** [openthread/openthread#10761] [Q&A]
As we understood, Power calibration feature enabled calibration of power and support to set power per channel.

I have few queries on set target power per channel:

1) otPlatRadioSetChannelTargetPower() is similar to otPlatRadioSetTransmitPower() API only difference is otPlatRadioSetChannelTargetPower() set power per channel by reading configuration file from host and send to RCP while  otPlatRadioSetTransmitPower()  set power for current channel (assuming for all channel even if switch to different one) using ot-ctl command from host to RCP. Is it correct?

2) Can I use otPlatRadioSetChannelMaxTransmitPower() to define maximum limit of power per channel along with otPlatRadioSetChannelTargetPower() API? 

3) Feature mentioned otPlatRadioSetTransmitPower() API should be disabled if otPlatRadioSetChannelTargetPower()  is defined in RCP. 
May I know why it needs to be disabled?  Why we can't maintain both feature?  otPlatRadioSetTransmitPower() can configured power while host is running. otPlatRadioSetChannelTargetPower()  only set power per channel when host is initialized stack.  To update power table, we need to stop host and again required to re-run. 

4) Why we can't provide new CLI to update power per channel using ot-ctl as well ? or can't we update existing ot-ctl txpower CLI to update power per channel and completely  obsolete otPlatRadioSetTransmitPower() API?

**Comment by zhanglongxia** [comment]
1. Correct. The `otPlatRadioSetChannelTargetPower()` is used to set the power for a given channel. The `otPlatRadioSetTransmitPower ()` will set all channels to use the same power.
2. For historical reasons, we have these two similar functions. Function `otPlatRadioSetChannelTargetPower` can provide a higher precision power settings. I recommend using the `otPlatRadioSetChannelTargetPower()` to replace `otPlatRadioSetChannelMaxTransmitPower `.
3. Use `otPlatRadioSetChannelTargetPower()` and `otPlatRadioSetTransmitPower()` at the same time will cause displaying wrong power value. For example, the user sets the power to a fixed value via `otPlatRadioSetTransmitPower()`, then an app calls the `otPlatRadioSetChannelTargetPower()` to change the power of another channel. The cli command still show the power value set by `otPlatRadioSetTransmitPower`, this power value is different with the actual set power value in Thread radio chip. It will be very confusing. So we suggest not enabling them both at the same time. If you are not care about this issue, you can implement these function at the same time.  Note: `otPlatRadioSetChannelTargetPower()` also could be called in non-initial stage.
4. It should be fine to add a new cli to set power per channel.

**Reply by aashu216** [reply]
for point 2, I am in impression otPlatRadioSetChannelTargetPower () is set power per channel and otPlatRadioSetChannelMaxTransmitPower () is set maximum limit of power per channel. 

So, power set by otPlatRadioSetChannelTargetPower () should not exceed power set by otPlatRadioSetChannelMaxTransmitPower (). correct?

**Reply by aashu216** [reply]
@zhanglongxia, can you please provide your review?

**Reply by aashu216** [reply]
Can you help to clear above doubts?

**Reply by zhanglongxia** [reply]
@aashu216 Sorry for the late reply. Both of these two functions are optional platform functions (return OT_ERROR_NOT_IMPLEMENTED). For historical reasons, we retain two functions. When we actually use them in products, we will only choose to implement one of the functions. We usually implement the function `otPlatRadioSetChannelTargetPower`, because its definition is more accurate, the power setting accuracy is higher.

**Reply by aashu216** [reply]
Thanks for response but still I need some clearance.  otPlatRadioSetChannelTargetPower () is set maximum limit of power specific to channel or it set the power used by radio?

**Reply by jwhui** [reply]
I believe this is documented here:

https://github.com/openthread/openthread/blob/722b15059aa45f31e2cbf18d01a241c7ebb31722/include/openthread/platform/radio.h#L301-L317

**Reply by aashu216** [reply]
Thanks @jwhui , 
Although so all functions are used to set maximum power limit then why here it is mentioned to disable otPlatRadioSetTransmitPower() (while it is setting power instead of limit) when otPlatRadioSetChannelTargetPower() is implemtened?

![image](https://github.com/user-attachments/assets/364cb6fe-9075-497d-8aa9-83851b43fb6f)

**Reply by jwhui** [reply]
It's a "should", as in a recommendation. Note that OpenThread does not actually make use of these functions. It's up to the platform / product to use them as appropriate.

**Reply by zhanglongxia** [reply]
> Thanks for response but still I need some clearance. otPlatRadioSetChannelTargetPower () is set maximum limit of power specific to channel or it set the power used by radio?

The `otPlatRadioSetChannelTargetPower()` sets the power used by radio.  Due to hardware limitations, Thread radio chip cannot output any configured power value. For example, user sets the output power to `7.5 dBm`, the thread radio chip can only output [`7dBm` or `8dBm`](https://github.com/openthread/ot-nrf528xx/blob/main/third_party/NordicSemiconductor/drivers/radio/nrf_802154_pib.c#L99). Based on the [description](https://github.com/zhanglongxia/openthread/blob/master/include/openthread/platform/radio.h#L1383) of the function `otPlatRadioSetChannelTargetPower`, the radio driver should output `7dBm`.

**Reply by aashu216** [reply]
Thanks @jwhui @zhanglongxia ,
can target power limit set in negative value (i.e., -5 dBm)?

**Reply by jwhui** [reply]
Yes, of course. The API specifies the type as a signed integer.

**Reply by aashu216** [reply]
Thanks @jwhui , @zhanglongxia 

for example, I have set power limit to 20 dBm for channel 11 using otPlatRadioSetChannelTargetPower() and now I have configured power level 18 dBm,  Radio should send packet on 18 dBm power in range but if I configure power level 21 dBm, Radio will set power level to 20 dBm correct?

And as I know I think no any regulatory region supports negative TX power limit so why API has support for negative limit value?

**Reply by jwhui** [reply]
The API may be used to limit transmit power for other reasons, including testing.

We also designed the API to support future compatibility with other RF bands and regions.

**Reply by zhanglongxia** [reply]
> for example, I have set power limit to 20 dBm for channel 11 using otPlatRadioSetChannelTargetPower() and now I have configured power level 18 dBm, Radio should send packet on 18 dBm power in range but if I configure power level 21 dBm, Radio will set power level to 20 dBm correct?
> 

The current description of the API `otPlatRadioSetChannelTargetPower()` and `otPlatRadioGetTransmitPower()` doesn't define how to process this case. It depends on how the platform implements them.

**Comment by zhanglongxia** [comment]
Refer to the latest update of the platform power related APIs: https://github.com/openthread/openthread/pull/10839

**Reply by aashu216** [reply]
Thanks @zhanglongxia
